### ISSUE TITLE
Strict error checking breaks custom VMs

### ISSUE THREAD

**Issue by smartcontracts** [nomiclabs/hardhat#1317]
I have a feeling this is going to be a relatively niche issue, sorry in advance for that! Optimism's hardhat plugin makes use of a custom ethereumjs-vm fork that gets injected into hardhat and replaces the nomiclabs/ethereumjs-vm instance. It's a bit of a hack but it's worked surprisingly well. 

A few days ago we started seeing issues with hardhat tooling. @bobface did a great analysis and pinned it down to this change from about a week ago: https://github.com/nomiclabs/hardhat/blob/3be60c05dc7bd7c37da3a38ba64c62e75fade831/packages/hardhat-core/src/internal/hardhat-network/provider/node.ts#L1166-L1168

The custom ethereumjs-vm instance we used is throwing `VmError` errors, but technically isn't throwing hardhat's `VmError` errors. So the instanceof check fails and the `VmError` is thrown as an internal error (which ends up breaking everything down the line). 

Would it be possible to replace this check with a slightly weaker that accepts anything that looks like a `VmError`?

**Comment by alcuadrado** [comment]
This is really interesting! I knew you used a custom ethereumjs-vm, but I thought you were adding a new provider. Does replacing it inside Hardhat Network mean that you get stack traces and console.log?

Good catch, @Bobface! Debugging this `instanceof` weirdness is always painful. Do you think this check would be ok?

```typescript
    if (
      !("error" in error) ||
      typeof error.error !== "string" ||
      error.error === ERROR.INTERNAL_ERROR
    ) {
      throw error;
    }
```

**Comment by smartcontracts** [comment]
> Does replacing it inside Hardhat Network mean that you get stack traces and console.log?

Yep, exactly :-) works really well out of the box!

**Comment by fvictorio** [comment]
Hey folks, we published this fix as `hardhat@rc`, could you please give it a try and let us know if it works?

**Comment by Bobface** [comment]
@fvictorio Yes works!

**Comment by alcuadrado** [comment]
> > Does replacing it inside Hardhat Network mean that you get stack traces and console.log?
> 
> Yep, exactly :-) works really well out of the box!

that's really clever 🤯👏

@Bobface, thanks for testing it. We'll release it as `latest` today.

**Comment by viraj124** [comment]
Thanks, Guys for the fix was struggling with this last week ❤️

**Comment by viraj124** [comment]
@alcuadrado can you post the package details to be updated for this issue ?

**Comment by viraj124** [comment]
got it never mind

**Comment by marvinkome** [comment]
I'm guessing this isn't published yet

**Comment by alcuadrado** [comment]
Oops, I think we messed something up and only released it as `hardhat@rc`. I'm reopening this until we make the actual release.

**Comment by fvictorio** [comment]
This should now be available in `hardhat@2.1.2`

**Comment by smartcontracts** [comment]
Nice! I'll go ahead and close this now. Thank you again :-)
